### PR TITLE
Tag Class and Related Test 

### PR DESCRIPTION
**Pull request by sbeesmiley**
placed constants within tag class and gave similar cases same message in is_tag_name_valid function
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
few more changes to clean up is_valid_tag_name tests based on feedback
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
is_valid_tag_name function changed to a static method and its corresponding tests were re-formatted for better readability, as suggested. 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Set min and max lengths as constants, is_valid_function returns a tuple with specific messages depending on the input and the tests for is_valid_tag_name was updated to check for additional cases and return messages.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Set min length requirement, streamlined format_tag function and called it in constructor, added one more catch according to comments.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Renamed set_tag function as recommended to format_tag
Added an if statement to catch empty string in is_valid_tag_name function as recommended
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Updated PR for CFM_65_SS
Made changes based on comments 
All tests passed for me locally. 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Updated PR for CFM_65_SS: 
followed the suggested style changes and added more tests as requested. 
All tests passed for me locally. 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
This PR is for CFM_65_SS. 
I added a Tag class to db.py and a test in the unit folder.

### CODE REVIEW

**Comment by CollectiveFocus** [comment]
The code looks good to me :). Going to test locally before approval

File: CommunityFridgeMapApi/dependencies/python/db.py [hunk 1]
```diff
@@ -198,11 +198,47 @@ def __init__(self, db_client: 'botocore.client.DynamoDB'):
...
 
-class Tag(DB_Item):
 
-    def __init__(self, db_client: 'botocore.client.DynamoDB'):
+class Tag(DB_Item):
+    TABLE_NAME = "tag"
+    def __init__(self, db_client: 'botocore.client.DynamoDB', tag_name:str= None):
         super().__init__(db_client=db_client)
+
+        self.db_client = db_client
```
Comment (CollectiveFocus) [inline CommunityFridgeMapApi/dependencies/python/db.py:211]: don't need to set db_client because it's already set in DB_Item

File: CommunityFridgeMapApi/dependencies/python/db.py [hunk 2]
```diff
@@ -198,11 +198,47 @@ def __init__(self, db_client: 'botocore.client.DynamoDB'):
...
+        self.tag_name = tag_name
+
+    def set_tag(self, tag_name):
```
Comment (CollectiveFocus) [inline CommunityFridgeMapApi/dependencies/python/db.py:214]: missing type hints here

File: CommunityFridgeMapApi/dependencies/python/db.py [hunk 3]
```diff
@@ -198,11 +198,47 @@ def __init__(self, db_client: 'botocore.client.DynamoDB'):
...
+            return True
+
+    def is_valid_tag_name(self, tag_name) --> bool:
+        #a valid tag name is alphanumeric, can include hyphen, underscore
+        for x in tag_name:
+            if not x.isalnum() and x not in ['_', '-']:
+                return False
+        return True
+
+    def add_item(self) -> DB_Response:
```
Comment (CollectiveFocus) [inline CommunityFridgeMapApi/dependencies/python/db.py:230]: add_item is missing unit test

Comment (sbeesmiley) [reply CommunityFridgeMapApi/dependencies/python/db.py:230]: > add_item is missing unit test

okay! I wasn't sure if I needed to write one for that as well. Do I also need to write one for set_tag? I will add to the test_tag in units/ tests. Thank you

Comment (CollectiveFocus) [reply CommunityFridgeMapApi/dependencies/python/db.py:230]: yes, lets write one for set_tag too

File: CommunityFridgeMapApi/dependencies/python/db.py [hunk 4]
```diff
@@ -198,11 +198,47 @@ def __init__(self, db_client: 'botocore.client.DynamoDB'):
...
+        is_valid_field = self.is_valid_tag_name(self.tag_name)
+        if not is_valid_field:
+            return DB_Response(message = "Tag Name Can Only Contain Letters, Numbers, Hyphens and Underscore: %s" % self.tag_name, status_code=400, success=False)
```
Comment (CollectiveFocus) [inline CommunityFridgeMapApi/dependencies/python/db.py:233]: style - this line is a bit long, can we make it multi-line?

DB_Resonse(
    message="",
    status_code=...
)

File: CommunityFridgeMapApi/dependencies/python/db.py (hunk 3 above)
Comment (CollectiveFocus) [reply CommunityFridgeMapApi/dependencies/python/db.py:230]: can we add a unit test for if the tag_name is None, we are not currently handling that case gracefully. I believe an exception will be thrown at the moment

File: CommunityFridgeMapApi/tests/unit/test_tag_is_valid.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
+from multiprocessing import connection
```
Comment (CollectiveFocus) [inline CommunityFridgeMapApi/tests/unit/test_tag_is_valid.py:1]: I think we can delete this file? Since you a already test `is_valid_tag_name` in the test_tag file

File: CommunityFridgeMapApi/dependencies/python/db.py [hunk 6]
```diff
@@ -198,11 +198,51 @@ def __init__(self, db_client: 'botocore.client.DynamoDB'):
...
 
-    def __init__(self, db_client: 'botocore.client.DynamoDB'):
+class Tag(DB_Item):
+    REQUIRED_FIELDS = ['tag_name']
+    TABLE_NAME = "tag"
+    def __init__(self, db_client: 'botocore.client.DynamoDB', tag_name:str= None):
         super().__init__(db_client=db_client)
+        self.tag_name = tag_name
+
+    def set_tag(self, tag_name) -> str:
```
Comment (CollectiveFocus) [inline CommunityFridgeMapApi/dependencies/python/db.py:213]: type hint missing on tag_name

should be  ```set_tag(self, tag_name:str) -> str:```

File: CommunityFridgeMapApi/tests/unit/test_tag.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+from dependencies.python.db import get_ddb_connection
+import unittest
+from dependencies.python.db import Tag
+
+def test_layer_test():
+
+    ret = layer_test()
+    assert ret == "hello world"
+
+def test_get_ddb_connection():
```
Comment (CollectiveFocus) [inline CommunityFridgeMapApi/tests/unit/test_tag.py:13]: You don't have to test this function because it's already tested in the Fridge test file I believe

File: CommunityFridgeMapApi/tests/unit/test_tag.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
+from multiprocessing import connection
+
+from dependencies.python.db import layer_test
+from dependencies.python.db import get_ddb_connection
+import unittest
+from dependencies.python.db import Tag
+
+def test_layer_test():
```
Comment (CollectiveFocus) [inline CommunityFridgeMapApi/tests/unit/test_tag.py:8]: don't have to test this, it's already tested in the `test_fridge.py` file

I should probably not have the test there to be honest! I'll move it to a different test file at some point :)

File: CommunityFridgeMapApi/tests/unit/test_tag_is_valid.py (hunk 5 above)
Comment (sbeesmiley) [reply CommunityFridgeMapApi/tests/unit/test_tag_is_valid.py:1]: I deleted that from my branch before the commit and push. I'll check again!

File: CommunityFridgeMapApi/tests/unit/test_tag_is_valid.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+import unittest
+from dependencies.python.db import Tag
+
+def test_layer_test():
+
+    ret = layer_test()
+    assert ret == "hello world"
+
+def test_get_ddb_connection():
+    connection = get_ddb_connection()
```
[the remark this answers is not in the document]
Comment (sbeesmiley) [reply CommunityFridgeMapApi/tests/unit/test_tag_is_valid.py]: This is my initial file which has been deleted. It wasn't deleted immediately, (sorry!), so that's probably why it showed up for you.  In the recent commit, this test is deleted so it shouldn't be there now for you. It is still  included in test_fridge.py where I had copied it from.

File: CommunityFridgeMapApi/dependencies/python/db.py [hunk 10]
```diff
@@ -198,11 +198,47 @@ def __init__(self, db_client: 'botocore.client.DynamoDB'):
...
+    TABLE_NAME = "tag"
+    def __init__(self, db_client: 'botocore.client.DynamoDB', tag_name:str= None):
         super().__init__(db_client=db_client)
+
+        self.db_client = db_client
+        self.tag_name = tag_name
+
+    def set_tag(self, tag_name):
+        #Tag_name is alphanumeric, can include hyphen and underscore, with no spaces and all lower cased
+        tag_name = tag_name.lower().replace(" ", "")
```
[the remark this answers is not in the document]
Comment (sbeesmiley) [reply CommunityFridgeMapApi/dependencies/python/db.py:215]: The thought process for me for this function is the tag_name may include spaces, capital letters, or other unacceptable characters; so, this function "sets" it in the appropriate form by removing the spaces, making it lower case, if needed, and checking that each character is acceptable. The latest commit does have the type hint for string but not a bool.  Regarding a crash, the test file does test a tag_name  with a None value and the test passes. Do you think additional tests are needed?

File: CommunityFridgeMapApi/dependencies/python/db.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sbeesmiley) [reply CommunityFridgeMapApi/dependencies/python/db.py]: I combined this comment in the above- I can add a bool type hint or remove the return value. Which do you think is best?

File: CommunityFridgeMapApi/dependencies/python/db.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (abhinav1912) [reply CommunityFridgeMapApi/dependencies/python/db.py:215]: 1. Should we rename it to `format_tag()` then? Since we're just formatting something that was already set
2. I'm not talking about the tests. At runtime, `tag_name.lower()` will crash if `tag_name` is `None`. So either we need to check if `tag_name` is non-null, or handle the `None` value as well

File: CommunityFridgeMapApi/dependencies/python/db.py [hunk 11]
```diff
@@ -198,11 +198,51 @@ def __init__(self, db_client: 'botocore.client.DynamoDB'):
...
+    def set_tag(self, tag_name:str) -> str:
+        #Tag_name is alphanumeric, can include hyphen and underscore, with no spaces and all lower cased
+        tag_name = tag_name.lower().replace(" ", "")
+        if not self.is_valid_tag_name(tag_name):
+            return False
+        else:
+            self.tag_name = tag_name
+            return True
+
+    def is_valid_tag_name(self, tag_name:str) -> bool:
```
Comment (abhinav1912) [inline CommunityFridgeMapApi/dependencies/python/db.py:222]: Would an empty string be a valid value?

**Review by abhinav1912 — APPROVED** [review]

File: CommunityFridgeMapApi/dependencies/python/db.py [hunk 12]
```diff
@@ -198,11 +198,58 @@ def __init__(self, db_client: 'botocore.client.DynamoDB'):
...
 
-    def __init__(self, db_client: 'botocore.client.DynamoDB'):
+class Tag(DB_Item):
+    REQUIRED_FIELDS = ['tag_name']
+    TABLE_NAME = "tag"
+    def __init__(self, db_client: 'botocore.client.DynamoDB', tag_name:str= None):
         super().__init__(db_client=db_client)
+        self.tag_name = tag_name
+
+    def format_tag(self, tag_name:str) -> str:
```
Comment (CollectiveFocus) [inline CommunityFridgeMapApi/dependencies/python/db.py:213]: We never actually use this function? 

And I think this function might be doing too much?

```
def format_tag(self, tag_name:str) -> str:
    if tag_name:
        tag_name = tag_name.lower().replace(" ", "")
    return tag_name
```

And you can then call this function within the constructor

```
self.tag_name = format_tag(tag_name)
```

File: CommunityFridgeMapApi/dependencies/python/db.py [hunk 13]
```diff
@@ -198,11 +198,58 @@ def __init__(self, db_client: 'botocore.client.DynamoDB'):
...
+            message = "Tag Name Can Only Contain Letters, Numbers, Hyphens and Underscore: %s" % self.tag_name,
+            status_code=400, success=False)
+        item = {"tag_name" : { "S": self.tag_name }}
+
+        try:
+            self.db_client.put_item(
+                TableName=self.TABLE_NAME,
+                Item=item
+            )
+        except ClientError as e:
```
Comment (CollectiveFocus) [inline CommunityFridgeMapApi/dependencies/python/db.py:252]: Need to add one more catch: 

```
        except self.db_client.exceptions.ResourceNotFoundException as e:
            message = "Cannot do operations on a non-existent table:  %s" % Tag.TABLE_NAME
            logging.error(message)
```

**Review by CollectiveFocus — CHANGES_REQUESTED** [review]
Made some small comments

File: CommunityFridgeMapApi/dependencies/python/db.py [hunk 14]
```diff
@@ -191,18 +191,60 @@ def add_item(self) -> DB_Response:
...
+            return False
+
+    def add_item(self) -> DB_Response:
+        has_required_fields, field = self.has_required_fields()
+        if not has_required_fields:
+            return DB_Response(message = "Missing Required Field: %s" % field, status_code=400, success=False)
+        is_valid_field = self.is_valid_tag_name(self.tag_name)
+        if not is_valid_field:
+            return DB_Response(
+            message = "Tag Name Can Only Contain Letters, Numbers, Hyphens and Underscore: %s" % self.tag_name,
```
Comment (CollectiveFocus) [inline CommunityFridgeMapApi/dependencies/python/db.py:234]: We should probably set a max length too. What's a good max length for a tag do you think?

And also we should return a different error message if the length of the tag is the issue.  A good way to do this might be if `is_valid_tag_name` returns a tuple instead with a boolean and an error message. Where the error message is an empty string if it is a valid tag name. 

Side Note:

This is something that will probably be done in multiple classes so we could abstract it out. But probably not worth to worry about it right now. But something good to look up for fun would be how other projects do request validation on lambdas that use api gateway. I think api gateway has a form of request validation that takes the burden away from the developers, but I could be wrong

File: CommunityFridgeMapApi/dependencies/python/db.py [hunk 15]
```diff
@@ -191,48 +191,39 @@ def add_item(self) -> DB_Response:
...
-                return True
-        else:
-            return False
+        return tag_name
 
     def is_valid_tag_name(self, tag_name:str) -> bool:
-        #A valid tag name is alphanumeric, can include hyphen, underscore
-        if tag_name:
+        #A valid tag name is alphanumeric, can include hyphen, underscore, with no spaces and all lower cased
+        if tag_name and len(tag_name) >= 3:
```
Comment (abhinav1912) [inline CommunityFridgeMapApi/dependencies/python/db.py:219]: nit: min tag length should be defined as a constant imo

File: CommunityFridgeMapApi/dependencies/python/db.py (hunk 14 above)
Comment (abhinav1912) [reply CommunityFridgeMapApi/dependencies/python/db.py:234]: An alternative to returning a tuple would be to raise an exception and catch it in the calling function

File: CommunityFridgeMapApi/dependencies/python/db.py (hunk 15 above)
Comment (CollectiveFocus) [reply CommunityFridgeMapApi/dependencies/python/db.py:219]: I agree

File: CommunityFridgeMapApi/tests/unit/test_tag.py [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+class DynamoDbMockPutItem():
+    def __init__(self):
+        pass
+
+    def put_item(self, TableName=None, Item=None, ConditionExpression=None):
+        pass
+
+class TagTest(unittest.TestCase):
+
+    def test_is_valid_tag_name(self):
```
Comment (abhinav1912) [inline CommunityFridgeMapApi/tests/unit/test_tag.py:13]: Can we refactor this test slightly for easier modification and readability? Something like
```
tag = Tag(…)
expected_false = [tag.is_valid_tag_name('')[0], tag.is_valid_tag_name(None)[0], …]
expected_equal = [(tag.is_valid_tag_name('')[1], 'tag_name is an empty string'), (tag.is_valid_tag_name(None)[1], 'tag_name is None'), …]
for condition in expected_false:
    self.assertFalse(condition)
for condition in expected_equal:
    self.assertEqual(condition)
```

File: CommunityFridgeMapApi/dependencies/python/db.py [hunk 17]
```diff
@@ -293,5 +294,62 @@ def __init__(self, db_client: "botocore.client.DynamoDB"):
...
         super().__init__(db_client=db_client)
+        self.tag_name = self.format_tag(tag_name)
+
+    def format_tag(self, tag_name:str) -> str:
+        #Tag_name is alphanumeric, can include hyphen and underscore, with no spaces and all lower cased
+        if tag_name:
+            tag_name = tag_name.lower().replace(" ", "")
+        return tag_name
+
+    def is_valid_tag_name(self, tag_name:str) -> Tuple[bool, str]:
```
Comment (abhinav1912) [inline CommunityFridgeMapApi/dependencies/python/db.py:310]: This method should belong to the class, since it isn’t instance specific. Can we make this a static method on the Tag class?

File: CommunityFridgeMapApi/tests/unit/test_tag.py [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+        tag = Tag(tag_name='test-tag', db_client=None)
+        self.assertTrue(tag.is_valid_tag_name('test-tag')[0])
+        self.assertEqual(tag.is_valid_tag_name('test-tag')[1], '')
+        tag = [
```
Comment (abhinav1912) [inline CommunityFridgeMapApi/tests/unit/test_tag.py:17]: This variable isn’t required

File: CommunityFridgeMapApi/tests/unit/test_tag.py [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+            Tag(tag_name='', db_client=None),
+            Tag(tag_name=None, db_client=None),
+            Tag(tag_name='ab', db_client=None),
+            Tag(tag_name='test&tag', db_client=None),
+            Tag(tag_name='TEST-tag_tag123xyui1234567890__--', db_client=None)
+        ]
+        expected_false = [
+            tag[0].is_valid_tag_name('')[0],
```
Comment (abhinav1912) [inline CommunityFridgeMapApi/tests/unit/test_tag.py:25]: You can simply use `Tag.is_valid…` since this is a static method

File: CommunityFridgeMapApi/tests/unit/test_tag.py [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+            (tag[0].is_valid_tag_name('')[1], 'tag_name is an empty string'),
+            (tag[1].is_valid_tag_name(None)[1], 'tag_name is None'),
+            (tag[2].is_valid_tag_name('ab')[1], 'Length of the tag_name is 2. It should be >= 3 but <= 32.'),
+            (tag[3].is_valid_tag_name('test&tag')[1], 'tag_name contains invalid characters'),
+            (tag[4].is_valid_tag_name('TEST-tag_tag123xyui1234567890__--')[1],
+                'Length of the tag_name is 33. It should be >= 3 but <= 32.')
+        ]
+        for i in range(5):
+            self.assertFalse(expected_false[i])
+            self.assertEqual(expected_equal[i][0], expected_equal[i][1])
```
Comment (abhinav1912) [inline CommunityFridgeMapApi/tests/unit/test_tag.py:41]: nit: `self.assertEqual(expected_equal[i])`

Comment (sbeesmiley) [reply CommunityFridgeMapApi/tests/unit/test_tag.py:41]: when I use this I get an Assertion Error stating that two arguments are required.

File: CommunityFridgeMapApi/tests/unit/test_tag.py (hunk 19 above)
Comment (abhinav1912) [reply CommunityFridgeMapApi/tests/unit/test_tag.py:25]: Why would lines get longer? You are replacing `tag[i]` with `Tag`

Comment (sbeesmiley) [reply CommunityFridgeMapApi/tests/unit/test_tag.py:25]: Okay- I like it that way. It looks cleaner.

File: CommunityFridgeMapApi/tests/unit/test_tag.py (hunk 20 above)
Comment (abhinav1912) [reply CommunityFridgeMapApi/tests/unit/test_tag.py:41]: I’m reviewing this on my tablet, so can’t see the entire codebase. Ideally `self.assertEqual` should also be able to de-structure a tuple as well. But this should be acceptable for now.

File: CommunityFridgeMapApi/dependencies/python/db.py [hunk 21]
```diff
@@ -293,5 +294,76 @@ def __init__(self, db_client: "botocore.client.DynamoDB"):
...
+        self.tag_name = self.format_tag(tag_name)
+
+    def format_tag(self, tag_name:str) -> str:
+        #Tag_name is alphanumeric, can include hyphen and underscore, with no spaces and all lower cased
+        if tag_name:
+            tag_name = tag_name.lower().replace(" ", "")
+        return tag_name
+
+    @staticmethod
+    def is_valid_tag_name(tag_name:str) -> Tuple[bool, str]:
```
Comment (CollectiveFocus) [inline CommunityFridgeMapApi/dependencies/python/db.py:311]: I have a suggestion on this method:

* ``'tag_name is None'``, ```tag_name is an empty string```, and `Length of the tag_name is {length_tag_name}. It should be >= 3 but <= 32` can all be under the same error message. They all basically tell the user that the length of the tag name is incorrect. So I think error message `Length of the tag_name is {length_tag_name}. It should be >= 3 but <= 32` is enough. What do you think?
```
is_tag_length_valid = length_tag_name >= MIN_TAG_LENGTH and length_tag_name <= MAX_TAG_LENGTH
if tag_name is None or is_tag_length_valid:
  message = f'Length of the tag_name is {length_tag_name}. It should be >= 3 but <= 32.'
```

File: CommunityFridgeMapApi/dependencies/python/constants.py [hunk 22]
```diff
@@ -0,0 +1,3 @@
+#Tag Class constants
+MIN_TAG_LENGTH = 3
```
Comment (CollectiveFocus) [inline CommunityFridgeMapApi/dependencies/python/constants.py:2]: I think a constants file would be good if we had global variables that more than one class uses. But since this constant is only used in the Tag class then I think these could live within the Tag class. What do you think?

File: CommunityFridgeMapApi/dependencies/python/db.py (hunk 21 above)
Comment (sbeesmiley) [reply CommunityFridgeMapApi/dependencies/python/db.py:311]: Good thought about using the constants in the error message - I changed that. 

I changed the error messages (except for None) to be the f string message.  

None and empty string are different types in Python so it will throw an error if I try to get the length of a None type. So, that's why I had that separated out. I will try to include the is_tag_length_valid variable though in the code- I like that.

Comment (CollectiveFocus) [reply CommunityFridgeMapApi/dependencies/python/db.py:311]: I realized that so I got rid of `is_tag_length_valid` in the example code that I added

File: CommunityFridgeMapApi/dependencies/python/constants.py (hunk 22 above)
Comment (sbeesmiley) [reply CommunityFridgeMapApi/dependencies/python/constants.py:2]: Okay! Constants aren't used a lot in Python so I was just following the standard instructions. I agree that it seems silly to have a separate file for those two.

File: CommunityFridgeMapApi/dependencies/python/db.py (hunk 21 above)
Comment (CollectiveFocus) [reply CommunityFridgeMapApi/dependencies/python/db.py:311]: But up to you :). If you want to include a `None` message I would suggest to make the error message: `Missing Required Field: tag_name`. Though in practice it'll never get there because we check for the required_fields first and that error will be handled there first

**Review by CollectiveFocus — APPROVED** [review]
Looks great! 🥳
